### ISSUE TITLE
windows support with CUDA

### ISSUE THREAD

**Issue by BinWang0213** [arborx/ArborX#321]
I'm tring to compile ArborX in windows using visual studio 2019. 

I just compiled kokkos in Windows with CUDA (see https://github.com/kokkos/kokkos/issues/3062)

But I can not compile ARBORX_ENABLE_EXAMPLES with following errors:
```
1> CMake generation started for configuration: 'x64-Release'.
1> Command line: "cmd.exe" /c ""C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\CMake\bin\cmake.exe"  -G "Ninja" -DCMAKE_INSTALL_PREFIX:PATH="C:\Users\BinPc\Source\Repos\ArborX\out\install\x64-Release" -DKokkos_DIR:PATH="C:/Users/BinPc/source/repos/kokkos/out/install/x64-Release/lib/cmake/Kokkos" -DARBORX_ENABLE_EXAMPLES:BOOL="True" -DBoost_DIR:PATH="C:/local/boost_1_71_0" -DBoost_INCLUDE_DIR:PATH="C:/local/boost_1_71_0" -DBoost_PROGRAM_OPTIONS_LIBRARY_DEBUG:FILEPATH="C:/local/boost_1_71_0/lib64-msvc-14.2/boost_program_options-vc142-mt-gd-x64-1_71.lib" -DBoost_PROGRAM_OPTIONS_LIBRARY_RELEASE:FILEPATH="C:/local/boost_1_71_0/lib64-msvc-14.2/boost_program_options-vc142-mt-x64-1_71.lib" -DARBORX_ENABLE_TESTS:BOOL="False" -DBoost_LIBRARY_DIR_DEBUG:PATH="C:/local/boost_1_71_0/lib64-msvc-14.2" -DBoost_LIBRARY_DIR_RELEASE:PATH="C:/local/boost_1_71_0/lib64-msvc-14.2" -DCMAKE_CXX_COMPILER=nvcc -DCMAKE_CXX_FLAGS="-ccbin \"C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\" " -DCMAKE_LINKER="C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/Llvm/bin/lld-link.exe" -DCMAKE_AR="C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/Llvm/bin/llvm-ar.exe" -DCMAKE_BUILD_TYPE="RelWithDebInfo" -DCMAKE_MAKE_PROGRAM="C:\PROGRAM FILES (X86)\MICROSOFT VISUAL STUDIO\2019\COMMUNITY\COMMON7\IDE\COMMONEXTENSIONS\MICROSOFT\CMAKE\Ninja\ninja.exe" "C:\Users\BinPc\Source\Repos\ArborX" 2>&1"
1> Working directory: C:\Users\BinPc\Source\Repos\ArborX\out\build\x64-Release
1> [CMake] -- Enabled Kokkos devices: CUDA;SERIAL
1> [CMake] -- ArborX hash = 'dbce44b'
1> [CMake] CMake Warning at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.15/Modules/FindBoost.cmake:1125 (message):
1> [CMake]   New Boost version may have incorrect or missing dependencies and imported
1> [CMake]   targets
1> [CMake] Call Stack (most recent call first):
1> [CMake]   C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.15/Modules/FindBoost.cmake:1247 (_Boost_COMPONENT_DEPENDENCIES)
1> [CMake]   C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.15/Modules/FindBoost.cmake:1885 (_Boost_MISSING_DEPENDENCIES)
1> [CMake]   examples/CMakeLists.txt:14 (find_package)
1> [CMake] -- Configuring done
1> [CMake] CMake Error in examples/access_traits/CMakeLists.txt:
1> [CMake]   No known features for CXX compiler
1> [CMake] 
1> [CMake]   ""
1> [CMake] 
1> [CMake]   version .
1> [CMake] 
1> [CMake] 
1> [CMake] CMake Generate step failed.  Build files cannot be regenerated correctly.
1> [CMake] 
```

Do you have any experience in this?

**Comment by dalg24** [comment]
Does your compiler not support c++14?
This comes from https://github.com/arborx/ArborX/blob/d42a8f3f5466ee814cfc97f832b8a160285e534d/CMakeLists.txt#L11

**Comment by BinWang0213** [comment]
I used nvcc (CUDA 10.2) to compile the code. I believe it support c++ 14. 
I didn't familar with cmake. How can I set this manully?

**Comment by dalg24** [comment]
Try commenting out the line I pointed out

**Comment by BinWang0213** [comment]
As I can compile the code without enable the examples, I create a Visual Studio project manully.

1. Create a CUDA project via File->New-Project-> CUDA 10.2 Runtime
2. Copy `example_cuda_access_traits.cpp` into `main.cu`
3. Set INCLUDE in `C/C++` path by append following path into `Additional Include Directories`:
    -C:\kokkos\include
    -C:\ArborX\install\include
    -C:\ArborX\install\include\details
4. Set LIBRARY in `Linker` by append following lib into `Additional Dependencies`:
    -C:\kokkos\lib\kokkoscore.lib
    -C:\kokkos\lib\kokkoscontainers.lib
5. `Build` the code

Here is the error I have:
```
1>------ Build started: Project: test, Configuration: Release x64 ------
1>Compiling CUDA source file kernel.cu...
1>
1>F:\nearest_neigh\ArborX\examples\access_traits\test>"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64" -x cu  -IF:\nearest_neigh\kokkos\include -IF:\nearest_neigh\ArborX\install\include -IF:\nearest_neigh\ArborX\install\include\details -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Release -maxrregcount=0  --machine 64 --compile -cudart static --expt-extended-lambda    -DWIN32 -DWIN64 -DNDEBUG -D_CONSOLE -D_MBCS -Xcompiler "/EHsc /W3 /nologo /O2 /Fdx64\Release\vc142.pdb /FS /Zi  /MD " -o x64\Release\kernel.cu.obj "F:\nearest_neigh\ArborX\examples\access_traits\test\kernel.cu"
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_Atomic_MinMax.hpp(117): error : no instance of overloaded function "atomicMin" matches the argument list
1>            argument types are: (unsigned long long *, const unsigned long long)
1>
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_Atomic_MinMax.hpp(133): error : no instance of overloaded function "atomicMax" matches the argument list
1>            argument types are: (unsigned long long *, const unsigned long long)
1>
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_Atomic_MinMax.hpp(196): error : no instance of overloaded function "atomicMin" matches the argument list
1>            argument types are: (unsigned long long *, const unsigned long long)
1>
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_Atomic_MinMax.hpp(215): error : no instance of overloaded function "atomicMax" matches the argument list
1>            argument types are: (unsigned long long *, const unsigned long long)
1>
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsKokkosExt.hpp(24): error : redefinition of default argument
1>
1>5 errors detected in the compilation of "C:/Users/PORESI~1/AppData/Local/Temp/tmpxft_00004b60_00000000-7_kernel.cpp1.ii".
1>kernel.cu
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\BuildCustomizations\CUDA 10.2.targets(764,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_30,code=\"sm_30,compute_30\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64" -x cu  -IF:\nearest_neigh\kokkos\include -IF:\nearest_neigh\ArborX\install\include -IF:\nearest_neigh\ArborX\install\include\details -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Release -maxrregcount=0  --machine 64 --compile -cudart static --expt-extended-lambda    -DWIN32 -DWIN64 -DNDEBUG -D_CONSOLE -D_MBCS -Xcompiler "/EHsc /W3 /nologo /O2 /Fdx64\Release\vc142.pdb /FS /Zi  /MD " -o x64\Release\kernel.cu.obj "F:\nearest_neigh\ArborX\examples\access_traits\test\kernel.cu"" exited with code 1.
1>Done building project "test.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

**Comment by dalg24** [comment]
This block was a workaround C++11 before we dropped support and needs to be removed.

What error do you get when you remove it?

**Comment by aprokop** [comment]
> What error do you get when you remove it?

Doesn't the fact that the block is triggered mean ArborX is being compiled with C++11 which we do not support?

**Comment by masterleinad** [comment]
Unless, ` /Zc:__cplusplus` is passed, `MSVC` always returns 199711L independent of what it actually supports (https://docs.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=vs-2019). Still, if `std::enable_if_t` makes problem after removing the compatibility block a `/std:c++14` flag might still be missing.

**Comment by BinWang0213** [comment]
This is the errors after remove and pass `/Zc:__cplusplus` and `/std:c++14`
https://github.com/arborx/ArborX/blob/d42a8f3f5466ee814cfc97f832b8a160285e534d/src/details/ArborX_DetailsKokkosExt.hpp#L21-L26


```
1>------ Build started: Project: test, Configuration: Release x64 ------
1>Compiling CUDA source file kernel.cu...
1>
1>F:\nearest_neigh\ArborX\examples\access_traits\test>"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_75,code=\"sm_75,compute_75\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64" -x cu  -IF:\nearest_neigh\kokkos\include -IF:\nearest_neigh\ArborX\install\include -IF:\nearest_neigh\ArborX\install\include\details -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Release -maxrregcount=0  --machine 64 --compile -cudart static --expt-extended-lambda    -DWIN32 -DWIN64 -DNDEBUG -D_CONSOLE -D_MBCS -Xcompiler "/EHsc /W3 /nologo /O2 /Fdx64\Release\vc142.pdb /FS /Zi  /MD " -o x64\Release\kernel.cu.obj "F:\nearest_neigh\ArborX\examples\access_traits\test\kernel.cu"
1>kernel.cu
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_HostThreadTeam.hpp(350): warning C4267: 'argument': conversion from 'size_t' to 'int', possible loss of data
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_HostThreadTeam.hpp(380): warning C4244: 'initializing': conversion from 'int64_t' to 'int', possible loss of data
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_HostThreadTeam.hpp(380): warning C4244: 'initializing': conversion from 'int64_t' to 'const int', possible loss of data
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_HostThreadTeam.hpp(387): warning C4244: 'initializing': conversion from 'int64_t' to 'int', possible loss of data
1>F:\nearest_neigh\kokkos\include\impl/Kokkos_HostThreadTeam.hpp(387): warning C4244: 'initializing': conversion from 'int64_t' to 'const int', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_Point.hpp(39): warning C4244: 'argument': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_Point.hpp(39): warning C4244: 'argument': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_Point.hpp(39): warning C4244: 'argument': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_Sphere.hpp(27): warning C4244: 'initializing': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsAlgorithms.hpp(187): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsAlgorithms.hpp(204): warning C4244: '=': conversion from 'double' to 'float', possible loss of data
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(172): note: see reference to class template instantiation 'ArborX::Details::InsertGenerator<PassTag,Predicates,Callback,OutputView,CountView,OffsetView,PermuteType>' being compiled
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(68): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(72): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(89): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(93): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(111): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(116): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(124): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(129): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(137): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(141): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2059: syntax error: ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2238: unexpected token(s) preceding ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2059: syntax error: '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2059: syntax error: '&&'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2988: unrecognizable template declaration/definition
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(155): error C2059: syntax error: ','
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBufferOptimization.hpp(159): error C2334: unexpected token(s) preceding '{'; skipping apparent function body
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2062: type 'unknown-type' unexpected
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2447: '{': missing function header (old-style formal list?)
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2065: 'Offset': undeclared identifier
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2923: 'Kokkos::is_view': 'Offset' is not a valid template type argument for parameter '<unnamed-symbol>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2888: 'Kokkos::is_view': symbol cannot be defined within namespace 'BoundingVolumeHierarchyImpl'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(179): error C2062: type 'void' unexpected
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(185): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(185): error C2447: '{': missing function header (old-style formal list?)
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2059: syntax error: '<end Parse>'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2143: syntax error: missing ';' before '{'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2143: syntax error: missing '>' before ';'
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2062: type 'unknown-type' unexpected
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2447: '{': missing function header (old-style formal list?)
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): error C2065: 'Offset': undeclared identifier
1>F:\nearest_neigh\ArborX\install\include\details\ArborX_DetailsBoundingVolumeHierarchyImpl.hpp(287): fatal error C1003: error count exceeds 100; stopping compilation
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Microsoft\VC\v160\BuildCustomizations\CUDA 10.2.targets(764,9): error MSB3721: The command ""C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\bin\nvcc.exe" -gencode=arch=compute_75,code=\"sm_75,compute_75\" --use-local-env -ccbin "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.26.28801\bin\HostX86\x64" -x cu  -IF:\nearest_neigh\kokkos\include -IF:\nearest_neigh\ArborX\install\include -IF:\nearest_neigh\ArborX\install\include\details -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include" -I"C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\include"     --keep-dir x64\Release -maxrregcount=0  --machine 64 --compile -cudart static --expt-extended-lambda    -DWIN32 -DWIN64 -DNDEBUG -D_CONSOLE -D_MBCS -Xcompiler "/EHsc /W3 /nologo /O2 /Fdx64\Release\vc142.pdb /FS /Zi  /MD " -o x64\Release\kernel.cu.obj "F:\nearest_neigh\ArborX\examples\access_traits\test\kernel.cu"" exited with code 2.
1>Done building project "test.vcxproj" -- FAILED.
```

It looks like the std::enable_if_t in MSVC is different from ArborX, this is where the error comes:
https://github.com/arborx/ArborX/blob/c500c049e68b0232ba35c646c40b3eb0d993a98e/src/details/ArborX_DetailsBufferOptimization.hpp#L68-L71

**Comment by masterleinad** [comment]
There is
```
nvcc warning : The -std=c++14 flag is not supported with the configured host compiler. Flag will be ignored.
```
Note that the correct flag for `MSVC` is `/std:c++14`. Can you make sure that `std::enable_if_t` works in a dummy example first with your setup?

**Comment by aprokop** [comment]
It seems that implicit conversion to boolean does not work properly. Could try to replace all `{}` with `::value` for `std::is_same`.

**Comment by BinWang0213** [comment]
@masterleinad I tested following code without any error:

```
#include <iostream>
#include <type_traits>

class foo
{
public:
  template <class S,
            typename std::enable_if<std::is_integral<S>::value, int>::type = 0>
  void bar(S &&s)
  {
    std::cout << "is integral\n";
  }
  template <class S,
            typename std::enable_if<!std::is_integral<S>::value, int>::type = 0>
  void bar(S &&s)
  {
    std::cout << "NOT integral\n";
  }
};

int main()
{
  foo f;
  f.bar(1);
  f.bar(2.0);
}
```

**Comment by masterleinad** [comment]
Right. Can you try to replace `typename std::enable_if<...>::type` with `std::enable_if_t<...>` in your code?
Does doing the change in the opposite direction in `ArborX` help? Have you tried @aprokop's suggestion?

**Comment by BinWang0213** [comment]
@aprokop @masterleinad Your suggestion works!

I replaced all `{}` with `::value` for `std::is_same` and `Kokkos::is_view`. 

Here is the log:
[log.txt](https://github.com/arborx/ArborX/files/4689603/log.txt)

**Comment by BinWang0213** [comment]
Do you have any plan to fix this WINDWOS compatbility issue in the future release? It's not easy for other winodws users to fix this manully.

**Comment by masterleinad** [comment]
We don't officially support Windows and don't test with it. Also, we only support building through `CMake`.

It would be good to understand the issue a little bit more, e.g. https://godbolt.org/z/o78AAN works just fine.

Just out of curiosity: What is your motivation for using `Windows`?

**Comment by BinWang0213** [comment]
Sure.  It will be nice to test more before change the code. 

BTW, I build ArborX using Cmake-gui in Windows. I compiled many other windows-compatible library using cmake as well. Kokkos is a special case that requires VS19 to generate compile command.

Personally, I developed my research code (particle simulation) in my windows workstation and use Visual Studio a lot. So I prefer to use ArborX under windows.
